ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java

```diff
@@ -837,6 +864,14 @@ public boolean save_state(OutputObjectState os, int t)
 		{
 			os.packInt(_heuristic);
 			os.packBoolean(_committed);
+
+			/*
+			 * Even though there's still a chance of crashing before saving _rolledBack,
+			 * it's still worth trying to reduce the chances of invoking rollback again
+			 * on the same XA resource
+			 */
+
+			os.packBoolean(_rolledBack);
```

Comment:
+1

Reply:
Sorry, now I have thought about this a bit more I wonder if it's worth putting this as a separate improvement? The reason I say this is because otherwise the object store will be incompatible when someone upgrades. Maybe as this is nice to have it could be moved to a "version 8" of Narayana?

Reply:
I'm not fully following why the OS would be incompatible when someone upgrades :-/

Reply:
If you imagine Narayana "7.0.2.Final" (let's say, but anything after this pull request were merged) were to be ran with an existing object store from a version prior then when doing a restore state of one of these objects it would not find a valid (as I suppose it would read some of the bytes of the XidImple [next thing to be written] as one) boolean to use and would then be basically in error. Please can you let me know if you would like me to expand on this?

Reply:
I think this looks like the last issue (to be clear, affecting both XAResourceRecord even if I resolved the other conversation) in this commit. If you agree and don't think it should break transaction integrity, I think an option to consider is to back out the save and restore capability of the XAResourceRecord and to consider to raise that as a separate JBTM for the next major release adding a release note to that enhancement that it will require users to use a fresh object store or something like that?

If you do think it would break transaction integrity to not save this, we need to consider including it further but I am not sure where to go with it then to not break existing applications. If we haven't already got this feature we could consider to introduce at the same time writing a version for record types into the object store for records in order to help with compatibility in the future but it won't help with breaking applications this time I suppose (except maybe it will because maybe we can somehow detect if they don't have a version we read in the old way)...

Reply:
> If you imagine Narayana "7.0.2.Final" (let's say, but anything after this pull request were merged) were to be ran with an existing object store from a version prior then when doing a restore state of one of these objects it would not find a valid (as I suppose it would read some of the bytes of the XidImple [next thing to be written] as one) boolean to use and would then be basically in error. Please can you let me know if you would like me to expand on this?

Thanks Tom, that clarifies it. On the other hand though, I don't see why a user would have txns in the OS during a Narayana upgrade. Maybe we should advice users to end/solve all txns before a Narayana upgrade (in our documentation)?

Reply:
If it's not already there, that would seem like an important thing to add to our documentation. I have added a JBTM to investigate this https://issues.redhat.com/browse/JBTM-3856

Reply:
Added a reply in that JBTM.

Reply:
Thank you, much appreciated!

Reply:
I agree with this being part of the fix. That said, it does break compatibility and only minimises a window.

I think as this is a known breaking change:
* JBTM-3856 should be considered a blocker to merging the change as is
* JBTM-3843 should be targeted for the next release that can have breaking changes. I presume that might be Narayana 8 or perhaps even Narayana 7.1 depending on what the project allows for
* If JBTM-3843 is backported to an earlier version of Narayana, it should be considered whether the breaking change part can be backed out as the inconvenience of a user dealing with a changed object store between versions does not seem to outweigh the benefit of preventing a spurious call to xa_rollback for this case

I think a useful enhancement to consider for Narayana would be that object store records are versioned somehow and designed to be forwards compatible such that if a later version of Narayana adds a new field, that version can still read the old format records if the changes would not affect data integrity.

Reply:
We removed the breaking change thus I think we can merge (and backport) this PR as is. @tomjenkinson, WDYT?

Reply:
I am going to resolve this thread. (and I definitely think we can backport the functional commit of the pull request, I would need to consider the benefit of backporting the commit which renamed the variables).

Thank you